Add userVisibleData, userVisibleDataFormat, and userNonVisibleData support to `.authenticate`

alepek opened the pull request:
These variables were already technically supported but were hard to use due to the types not containing them. I've also updated the tests to reflect this change, and verified the functionality with my own test BankID.

closes #69

alepek commented on test/manual/src/manualTests.ts:
```diff
@@ -1,4 +1,4 @@
-import { BankIdClient } from "bankid";
+import { BankIdClient } from "../../../";
```
Had to make this change for the test to run. Possibly my local env is misconfigured, or the package.json root config is?

pmmmwh replied:
These tests are suppose to run with a "linked" local BankID build, so let's revert.

```suggestion
import { BankIdClient } from "bankid";
```

pmmmwh commented on src/bankid.ts:
```diff
@@ -36,7 +38,6 @@ interface AuthOptionalRequirements {
 
 export interface SignRequest extends AuthRequest {
   userVisibleData: string;
-  userVisibleDataFormat?: "simpleMarkdownV1";
```
Since we're changing this - we might as well move `userNonVisibleData` to `AuthRequest` as well.

pmmmwh commented on test/manual/yarn.lock:
```diff
@@ -12,18 +12,19 @@ asynckit@^0.4.0:
 
 "bankid@file:../..":
-  version "3.1.2"
+  version "3.1.3"
```
This line we should keep.

pmmmwh commented on test/manual/yarn.lock:
```diff
@@ -12,18 +12,19 @@ asynckit@^0.4.0:
-  version "0.27.2"
-  resolved "https://registry.yarnpkg.com/axios/-/axios-0.27.2.tgz#207658cc8621606e586c85db4b41a750e756d972"
-  integrity sha512-t+yRIyySRTp/wua5xEr+z1q60QmLq8ABsS5O9Me1AsE5dfKqgnCFzwiCZZ/cGNd1lq4/7akDWMxdhVlucjmnOQ==
+axios@^1.6.4:
```
We've been putting off upgrading Axios due to potential breaking changes, so I would suggest we revert the changes here.

alepek replied:
I reverted all changes to the lockfile. Worth nothing is that the will reappear upon running `yarn` in the test project.

pmmmwh commented:
Thanks for the PR 😄

alepek commented:
Thanks for the feedback! Will address it later today.

pmmmwh commented on src/bankid.ts:
```diff
@@ -239,6 +240,26 @@ export class BankIdClient {
+      throw new Error("userVisibleDataFormat can only be simpleMarkdownV1.");
+    }
+
+    console.log(parameters);
+
```
```suggestion
```

pmmmwh commented on src/bankid.ts:
```diff
@@ -239,6 +240,26 @@ export class BankIdClient {
+        ? Buffer.from(parameters.userNonVisibleData).toString("base64")
+        : undefined,
+    };
+
+    console.log(parameters);
```
```suggestion
```

pmmmwh commented on test/manual/src/manualTests.ts:
```diff
@@ -17,37 +17,33 @@ async function main() {
     production: false,
   });
 
-  try {
```
I think it's nice to be able to run all of the tests here - so if auth fails we can still see if sign works - maybe what was missing is some way to track "errored" status?

alepek replied:
Fair enoguh! I can update it to produce a summary of some sort at the end.

alepek commented:
Thanks for the feedback and pardon the slightly sloppy remnants of my debugging. Will adjust according to feedback and get back to you.

alepek commented:
I've changed the source code according to feedback and updated the test suite to output success/failure messages for each separate test and a summary at the end. 

Example successful run
<img width="534" alt="Screenshot 2024-01-23 at 19 05 55" src="https://github.com/anyfin/bankid/assets/5012939/0ffa7db6-1ada-44e3-afed-0ce050b8bd6c">

Example failed run (manual cancel of first two tests)
<img width="522" alt="Screenshot 2024-01-23 at 19 06 54" src="https://github.com/anyfin/bankid/assets/5012939/e87d46ed-e799-42d1-9bd3-8d8100c43157">

pmmmwh commented on test/manual/src/manualTests.ts:
```diff
@@ -74,29 +80,40 @@ async function main() {
+  } else {
+    console.error(`❌ ${failedTests} test(s) failed`);
+  }
+});
```
Food for thought - should we keep a `.catch` here just in case? (Skimmed through the test code and I believe the only thing that can throw is client instantiation or `delay` 🤔 )

alepek replied:
Yes, that is a fair point. Committed your suggestion!

pmmmwh commented:
Looks great! I have only one more thought 🙇🏻

pmmmwh commented on test/manual/src/manualTests.ts:
```diff
@@ -74,29 +80,40 @@ async function main() {
-main()
-  .then(() => console.log("✅ All tests completed successfully"))
-  .catch(err => console.log(err));
+main().then(errors => {
+  const failedTests = errors.length;
+  if (failedTests === 0) {
+    console.log("✅ All tests completed successfully");
+  } else {
+    console.error(`❌ ${failedTests} test(s) failed`);
+  }
+});
```
```suggestion
main()
	.then(errors => {
	  const failedTests = errors.length;
	  if (failedTests === 0) {
	    console.log("✅ All tests completed successfully");
	  } else {
	    console.error(`❌ ${failedTests} test(s) failed`);
	  }
	})
	.catch(console.error);
```

pmmmwh reviewed (approved):
Thank you! I'll try to cut a release this week.
